pkg/syncer/storagepool/intended_state.go

```diff
@@ -407,14 +412,18 @@ func (c *SpController) updateVsanSnaIntendedState(ctx context.Context, vsanState
 	return hostCapacityErr
 }
 
-func (c *SpController) deleteIntendedState(ctx context.Context, spName string) bool {
-	for dsMoid, state := range c.intendedStateMap {
-		if state.spName == spName {
-			delete(c.intendedStateMap, dsMoid)
-			return true
+func (c *SpController) deleteIntendedState(ctx context.Context, spName string) (deleted bool) {
+	deleted = false
+	c.intendedStateMap.Range(func(key, value interface{}) bool {
```

Comment:
@bhavyachoudhary Can you confirm if the concurrent map write error was observed even after changing intendedState map to a Sync.Map (`intendedStateMap sync.Map`).
Just wanted to understand whether `Sync.Map` is not handling concurrent map write ?

Reply:
concurrent write error was not observed after changing intendedState map to a Sync.Map. 
Before making the changes these error logs were printed due to concurrent write to intendedStateMap variable
```
stderr F fatal error: concurrent map writes
stderr F
stderr F goroutine 25250 [running]:
stderr F runtime.throw(0x22885c4, 0x15)
stderr F         /build/mts/release/bora-17197275/compcache/cayman_go/ob-16783374/linux64/src/runtime/panic.go:1116 +0x72 fp=0xc0027d9658 sp=0xc0027d9628 pc=0x582b32
stderr F runtime.mapassign_faststr(0x1e870a0, 0xc0005d79b0, 0xc0019216e0, 0x29, 0xc0027d9818)
stderr F         /build/mts/release/bora-17197275/compcache/cayman_go/ob-16783374/linux64/src/runtime/map_faststr.go:211 +0x3f1 fp=0xc0027d96c0 sp=0xc0027d9658 pc=0x560c11
stderr F sigs.k8s.io/vsphere-csi-driver/pkg/syncer/storagepool.(*SpController).applyIntendedState(0xc00048e9a0, 0x25791a0, 0xc007744ab0, 0xc0044fd320, 0xc0027d9e22, 0x1b)
stderr F         /build/mts/release/bora-17197275/cayman_vsphere_csi_driver/vsphere_csi_driver/src/pkg/syncer/storagepool/intended_state.go:309 +0x3a5 fp=0xc0027d9a58 sp=0xc0027d96c0 pc=0x1b33405
stderr F sigs.k8s.io/vsphere-csi-driver/pkg/syncer/storagepool.(*SpController).updateVsanSnaIntendedState(0xc00048e9a0, 0x25791a0, 0xc007744ab0, 0xc004e64120, 0xc0027d9d30, 0xc00022c4d0, 0x0, 0x1)
stderr F         /build/mts/release/bora-17197275/cayman_vsphere_csi_driver/vsphere_csi_driver/src/pkg/syncer/storagepool/intended_state.go:380 +0x2e5 fp=0xc0027d9bb0 sp=0xc0027d9a58 pc=0x1b347c5
stderr F sigs.k8s.io/vsphere-csi-driver/pkg/syncer/storagepool.ReconcileAllStoragePools(0x25791a0, 0xc007744ab0, 0xc00022c4d0, 0xc00048e9a0, 0x0, 0x0)
stderr F         /build/mts/release/bora-17197275/cayman_vsphere_csi_driver/vsphere_csi_driver/src/pkg/syncer/storagepool/listener.go:237 +0x5f3 fp=0xc0027d9e50 sp=0xc0027d9bb0 pc=0x1b37ed3
stderr F sigs.k8s.io/vsphere-csi-driver/pkg/syncer/storagepool.scheduleReconcileAllStoragePools.func1(0xdf8475800, 0x5, 0xc000207f50, 0x25791a0, 0xc007744ab0, 0xc00022c4d0, 0xc00048e9a0)
stderr F         /build/mts/release/bora-17197275/cayman_vsphere_csi_driver/vsphere_csi_driver/src/pkg/syncer/storagepool/listener.go:180 +0x2ac fp=0xc0027d9fa8 sp=0xc0027d9e50 pc=0x1b4502c
stderr F runtime.goexit()
stderr F         /build/mts/release/bora-17197275/compcache/cayman_go/ob-16783374/linux64/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc0027d9fb0 sp=0xc0027d9fa8 pc=0x5b7fa1
stderr F created by sigs.k8s.io/vsphere-csi-driver/pkg/syncer/storagepool.scheduleReconcileAllStoragePools
stderr F         /build/mts/release/bora-17197275/cayman_vsphere_csi_driver/vsphere_csi_driver/src/pkg/syncer/storagepool/listener.go:172 +0xed
```

Reply:
Sounds good. And why do we need this deleted flag ? When Sync.Map is able to handle concurrent map write

Reply:
I have just rewritten the deleteIntendedState function using Sync.Map instead of regular map. If we want to iterate through Sync.Map, we cannot iterate like we do for regular map (`for dsMoid, state := range c.intendedStateMap`) we have to use Sync.Map's Range function. 
deleted flag is returned by the function and conveys whether or not the spName whose entry we wanted to remove was present in the  intendedStateMap.

Reply:
Thanks for confirming!